python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,189 @@
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        self.__class__._instances[self._ptr].add(self)
+        self._instances = self.__class__._instances[self._ptr]
+        self._zero_copied = False
+        weakref.finalize(self, _keys_cleanup, self._ptr)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+        exposed : bool, optional
+            Mark the buffer as zero copied.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self._instances) > 1
+
+    def get_ptr(self, mode: str = "write") -> int:
+        """Device pointer to the start of the buffer.
+
+        Parameters
+        ----------
+        mode : str, default 'write'
+            Supported values are {"read", "write"}
+            If "write", when weak-references exist, they
+            are unlinked and the data pointed to may be modified
+            by the caller. If "read", the data pointed to
+            must not be modified by the caller.
+            Failure to fulfill this contract will cause
+            incorrect behavior.
+
+        See Also
+        --------
+        Buffer.get_ptr
+        SpillableBuffer.get_ptr
+        """
+        if mode == "read":
+            return self._ptr
+        elif mode == "write":
+            self._unlink_shared_buffers()
+            return self._ptr
+        else:
+            raise ValueError(f"Incorrect mode passed : {mode}")
+
+    def copy(self, deep: bool = True):
+        if deep or self._zero_copied:
+            return super().copy(deep=True)
+        else:
+            cls = type(self)
+            copied_buf = cls.__new__(cls)
+            copied_buf._ptr = self._ptr
+            copied_buf._size = self._size
+            copied_buf._owner = self._owner
+            copied_buf._finalize_init()
+            return copied_buf
+
+    @property
+    def __cuda_array_interface__(self) -> dict:
+        # Unlink if there are any weak references.
+        self._unlink_shared_buffers()
```

Comment:
For clarity I would move this line below the comment just below. We wouldn't unlink if we weren't also setting to zero_copied, right? The two changes are coupled here for the same reason.

Reply:
Also, so that I understand, the implication is that accessing the CAI of a buffer that is currently sharing memory with other buffers will immediately trigger a copy, correct? i.e.
```
x = cudf.Series([1, 2, 3])
y = x[:2]
a = cp.asarray(x)  # This will force x to make a deep copy to unlink it from y
```

Reply:
> For clarity I would move this line below the comment just below. We wouldn't unlink if we weren't also setting to zero_copied, right? The two changes are coupled here for the same reason.

Right, Done 👍 



> Also, so that I understand, the implication is that accessing the CAI of a buffer that is currently sharing memory with other buffers will immediately trigger a copy, correct? i.e.
> 
> ```
> x = cudf.Series([1, 2, 3])
> y = x[:2]
> a = cp.asarray(x)  # This will force x to make a deep copy to unlink it from y
> ```

Yup, your understanding is correct.

Reply:
@galipremsagar unless my GH view isn't updated properly, it looks like the comments were merged but the unlink call is still happening above the comment. Not sure if that was intentional, I would still recommend moving that.